src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,181 @@
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    status = UCS_OK;
+    *event_set_p = event_set;
+    goto out_create;
+
+err_free:
+    ucs_free(event_set);
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+    if ( os_events == -1 ) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    memset(&event, 0, sizeof(event));
+    event.events   = os_events;
+    event.data.fd  = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_mod(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+    if ( os_events == -1 ) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    memset(&event, 0, sizeof(event));
+    event.events   = os_events;
+    event.data.fd  = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_MOD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, MOD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+{
+    int ret;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_DEL, event_fd, NULL);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, DEL, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+/*
+ *  event
+ *    epoll:  epoll_event.data.ptr
+ *    kqueue: kevent.udata
+ */
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+                                event_set_handler_t event_set_handler)
+{
+    int nready;
+    ucs_status_t status;
+    struct epoll_event events[UCS_EVENT_EPOLL_MAX_EVENTS];
+
+    nready = epoll_wait(event_set->epfd, events, UCS_EVENT_EPOLL_MAX_EVENTS,
+                        timeout_ms);
+
+    if ((nready < 0) && (errno != EINTR)) {
+        ucs_error("epoll_wait() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                   event_set->epfd, timeout_ms, nready);
+
+    status = event_set_handler(nready,events);
+    return status;
+}
+
+void ucs_event_set_cleanup(ucs_sys_event_set_t *event_set)
+{
```

Comment:
You said
> also, this function should return void

I'm misisng call `close()` descriptor.
Is this function still should return `void`?

Reply:
if close() fails it can print a warning
but the user doesn't really have anything to do with cleanup errors, similar reasoning to C++ destructor to not throw exceptions